### DISCUSSION TITLE
Is it possible to adapt Keccak to use 60-bit words?

### DISCUSSION THREAD

**Discussion by VictorTaelin** [XKCP/XKCP#98] [General]
[HVM](https://github.com/kindelia/hvm) has fast 60-bit native words, and slightly slower 64-bit boxed words. A close example would be Ocaml, which has unboxed 63-bit numbers, and only boxed 64-bit numbers. We're looking for the fastest hash function on HVM. Keccak is great, but it uses 64-bit words on its internal states, so, our best bet, so far, is to port it from implementations that use 32-bit numbers, such as JavaScript. That would be sub-optimal. I wonder if it would be possible to implement it using 60-bit words in its internal states, without having to emulate 64-bit words. It would be a different function, but that is fine. Is there any guideline on how to do so safely? Any reference implementation or pseudocode which is parametric on the word size?

**Comment by gvanas** [comment]
Keccak has a parametric lane size _w_, so the Keccak-_f_[1500] permutation with _w_=60 is in principle well-defined. However, we claim the security of instances of Keccak with _w_ a power of two only.

I wouldn't feel comfortable claiming security of Keccak-_f_[1500] without a least some further analysis. On the good side, the Keccak family has a Matryoshka structure (see Section 2.2 of [the Keccak reference](https://keccak.team/files/Keccak-reference-3.0.pdf)). Concretely, it means that the differential and linear trails in Keccak-_f_[100] (i.e., with _w_=4) extend symmetrically to Keccak-_f_[1500], although that is only a small part of the trails in Keccak-_f_[1500].

Before Keccak, we designed [RadioGatún](http://radiogatun.noekeon.org/), also with a parametric lane size. Here our claim on RadioGatún does include the case _w_=60. But it's a quite old design… :thinking:

**Comment by VictorTaelin** [comment]
I see, so I can just implement the same operations on U60. That sounds great; well done. I'll try. May I ask what further analysis would be required to establish its security, and what kind of budget would be necessary to complete it? I also wonder if it would be possible to decrease the number of rounds, to an extent, without affecting security in practice, in line with the *"Too Much Crypto"* argument.

**Comment by gvanas** [comment]
I would at least redo the trail analysis as done by [Mella et al., ToSC 2017](http://tosc.iacr.org/index.php/ToSC/article/view/597) to make sure there are no "accidents". This paper took several months of work; most likely just re-applying it to Keccak-_f_[1500] would take less time, but still it is by no means trivial.

And this is just one aspect of cryptanalysis next to [many others](https://keccak.team/third_party.html).

About the number of rounds, Keccak-_p_[1600, 12 rounds] is a good alternative (see [K12](https://github.com/XKCP/K12)), and it's safe in the light of Keccak's cryptanalysis track record. For Keccak-_f_[1500], 12 rounds could be enough too, to be seen.

Still, I am not sure I understand the benefits of switching to 60-bit words for performance reasons when the underlying hardware is 32- or 64-bit. Isn't the solution worse than the problem?

**Comment by gvanas** [comment]
Converting to a discussion, as it is not an issue in XKCP.

**Comment by VictorTaelin** [comment]
Thanks for the input, that is all the information I needed. Answering your question, the VM just doesn't have native unboxed 64-bit integers, like OCaml, so the options are to use boxed 64-bit integers (double memory) or emulated 64-bit integers (double memory and much less efficient). Just using 60-bit word size would make it 2x faster, at least, in such MV.

**Comment by VictorTaelin** [comment]
Just one last question, if we were to use Keccak-p[1600, 12 rounds], what is the right way to call it, and what do we cite? Is there a paper claiming its security?

**Reply by gvanas** [reply]
The name of this permutation is Keccak-_p_[1600, 12] and it is defined in Section 3.3 of [FIPS 202](https://csrc.nist.gov/publications/detail/fips/202/final).

There is no claim of security on the permutation itself. However, we claim it is secure for use in the sponge construction, as we did for K12. For this, you can [cite our ACNS 2018 paper](https://dblp.org/rec/conf/acns/BertoniDPAKV18.html).

**Reply by VictorTaelin** [reply]
Perfect, thank you. One last question: in a future, we might have 128-bit computer architectures in general. Would it be possible to optimize Keccak to that? For instance, perhaps using 3x3 lanes of 128-bits each? Or something along these lines?

**Reply by gvanas** [reply]
We could define a 128-bit variant of Keccak, but defining a new primitive requires further analysis and most preferably third-party cryptanalysis.

A simpler solution would be to use 128-bit words in a SIMD fashion to evaluate 2×Keccak-_p_[1600, 12] in parallel, which would be a perfect fit to K12. If your architecture does not support 2×64-bit rotations but only 128-bit rotations, it is always possible to interleave the bits of the two permutations inputs so as to replace 2×64-bit rotations with 128-bit rotations. (This is the opposite of the bit interleaving technique used to express 64-bit rotations as 2×32-bit rotations when implementing Keccak on 32-bit platforms.)
